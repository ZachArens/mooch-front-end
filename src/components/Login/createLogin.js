import React, { Component } from 'react';
import LoginForm from './loginForm';

//Call with props - updateFields function

export default class CreateLogin extends Component {

    render() {
        return (
            <div>
                <label>Verify email and password
                    <LoginForm emailName="verifyEmail" passwordName="verifyPassword" updateFields={this.props.updateFields} />
                </label>
                <div className="userDetails">
                    <label>Full Name
                        <input id="fullName" data-testid="fullName" name="fullName" type="text" onChange={this.props.updateFields}></input>
                    </label>
                    <label>Street Address
                        <input id="streetAddress" data-testid="streetAddress" name="streetAddress" type="text" onChange={this.props.updateFields}></input>
                    </label>
                    <label>City
                        <input id="city" data-testid="city" name="city" type="text" onChange={this.props.updateFields}></input>
                    </label>
                    <label>State (2 Letter)
                        <input id="st" data-testid="st" type="text" name="st" onChange={this.props.updateFields}></input>
                    </label>
                    <label>Zip
                        <input id="zip" data-testid="zip" type="text" name="zip" onChange={this.props.updateFields}></input>
                    </label>
                    <label>Phone
                        <input id="phone" data-testid="phone" type="text" name="phone" onChange={this.props.updateFields}></input>
                    </label>
                </div>
                
            </div>
        )
    }
}
