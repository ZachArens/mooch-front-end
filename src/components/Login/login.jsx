import React from 'react';
import { withRouter } from 'react-router-dom';
import {addUserDetails, createUserWithEmailandPass, loginWithEmailAndPass} from '../../utils/firebaseFunctions';
import SubmitButtons from '../submitButtons';
import LoginForm from './loginForm';
import CreateLogin from './createLogin';

//FIXME - uncaught error in login rejection
class Login extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            email: '', 
            password: '',
            verifyEmail: '',
            verifyPassword: '',
            fullName: '',
            streetAddress: '',
            city: '',
            st: '',
            zip: '',
            phone: '',
            errMsg: '',
            hasLogin: true,
            buttonText: "Don't have a login? Click here."
        };

        this.changeHasLogin = this.changeHasLogin.bind(this);
        this.submitEmailPass = this.submitEmailPass.bind(this);
        this.signUp = this.signUp.bind(this);
        this.updateFields = this.updateFields.bind(this);
    }

    updateFields = (e) => {
        const value = e.target.value;
        const name = e.target.name;

        this.setState({[name]: value, errMsg: ''});
        // console.log("update: " + name + " - " + value);
        
    }

    submitOrSignup = async (e) => {
        if(this.state.hasLogin) {
            await this.submitEmailPass(e);
        } else {
            await this.signUp(e);
        }

        const { history } = this.props;

        if (history) history.push(this.props.returnTo);
        
    }

    submitEmailPass = async (e) => {
        e.preventDefault();
        // console.log("email: " + this.state.email);
        // console.log("password: " + this.state.password);

        let userId = await loginWithEmailAndPass(this.state.email, this.state.password);

        // //FIXME - temp fix for auth problem
        // let userId = 'aQqcGAeDGafhOSQWeXDFA2klpuH2';
        this.props.setCurrentUser(userId);
    }

    signUp = async (e) => {
        e.preventDefault();

        //check if email and passwords match
        const emailIsMatch = this.state.email === this.state.verifyEmail;
        const passwordIsMatch = this.state.password === this.state.verifyPassword;

        if (!emailIsMatch) {
           this.setState({
               errMsg: "Your emails do not match. Please re-enter your email address and try again."
           });
           return null;
        } else if(!passwordIsMatch){
            this.setState({
                errMsg: "Your passwords do not match. Please re-enter your email address and try again."
            });
            return null;
        }

       let userId = createUserWithEmailandPass(this.state.email, this.state.password);

       this.props.setCurrentUser(userId);

       const userDetails = {
        uid: userId,
        email: this.state.email,
        fullName: this.state.fullName, 
        streetAddress: this.state.streetAddress,
        city: this.state.city, 
        st: this.state.st,
        zip: this.state.zip, 
        phone: this.state.phone,
       }

       console.log("adding user details: " + userDetails);

       addUserDetails(userDetails);
        
    };
    
    cancelFn = (e) => {
        this.setState = {email: '', password: '', errMsg: ''};
    }

    changeHasLogin = (e) => {
        e.preventDefault();
        let buttonText = this.state.hasLogin ? "Already have a login? Click here." : "Don't have a login? Click here.";
        
        this.setState((prevState) => ({
            hasLogin: !prevState.hasLogin,
            buttonText
        }));        

    }



    render() {

        // ui.start('#firebaseui-auth-container', uiConfig);

        // const haveOrCreateLogin = () => {
        //     if (this.state.hasLogin) {
        //         return <LoginForm updateFields={this.updateFields.bind(this)} />
        //     } else {
        //         return <CreateLogin updateFields={this.updateFields.bind(this)} />
        //     }
        // };

        return (
            <div data-testid="login">
                {this.state.errMsg && <div id="loginError" ><h4>{this.state.errMsg}</h4></div>}
                <form>
                    <title>Login</title>
                    <LoginForm emailName="email" passwordName="password" updateFields={this.updateFields} />
                    {!this.state.hasLogin && <CreateLogin updateFields={this.updateFields.bind(this)} />}
                    
                    <SubmitButtons submitTitle="Submit" cancelTitle="Cancel"
                        submitFn={this.submitOrSignup.bind(this)}
                        cancelFn={this.cancelFn.bind(this)} />

                    <button type="button" onClick={this.changeHasLogin} data-testid="loginOrCreateButton">{this.state.buttonText}</button>

                    {/*<div id="firebaseui-auth-container"/>*/}
                    {/*<div id="loader">Loading...</div>*/}

                </form>
                {/* <FireAuthWidget/> */}
            </div>

        );
    }
}

export default withRouter(Login);