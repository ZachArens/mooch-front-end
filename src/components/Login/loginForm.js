import React, { Component } from 'react';



export default class LoginForm extends Component {
    render() {
        return (
            <div>
                <label data-testid="emailLabel">Email
                    <input type="text" data-testid={this.props.emailName} placeholder="email" 
                    name={this.props.emailName} onChange={this.props.updateFields}/>
                </label>
                <label data-testid="passwordLabel">Password
                    <input type="password" data-testid={this.props.passwordName} 
                    name={this.props.passwordName} placeholder="********" onChange={this.props.updateFields}/>
                </label>
            </div>
        )
    }
}
