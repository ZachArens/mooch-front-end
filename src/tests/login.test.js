import React from 'react';
import ReactDOM from 'react-dom';
import Login from './../components/login';

import {render, cleanup} from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";

import renderer from "react-test-renderer";

afterEach(cleanup);

//tutorial provided by techsith at https://www.youtube.com/watch?v=3e1GHCA3GP0&t=4s

describe('renders Login component correctly', () => {
    it('renders without crashing', () => {
        const div = document.createElement("div");
        ReactDOM.render(<Login/>, div);
    });

    it('renders title correctly', () => {
        const {getByTestId} = render(<Login/>);
        expect(getByTestId('login')).toHaveTextContent("Login");
    });

    it('renders login form correctly', () => {
        const {getByTestId} = render(<Login/>);
        expect(getByTestId('login').getElementsByTagName('form').length).toBe(1);
    });

    it('renders email input correctly', () => {
        const {getByTestId} = render(<Login/>);
        const email = getByTestId('email');
        expect(email).toHaveTextContent("Email");
        expect(email.getElementsByTagName('input').length).toBe(1);
        expect(email.getElementsByTagName('input')[0].getAttribute('type')).toBe("text");
    });

    it('renders password input correctly', () => {
        const {getByTestId} = render(<Login/>);
        const password = getByTestId('password');
        expect(password).toHaveTextContent("Password");
        expect(password.getElementsByTagName('input').length).toBe(1);
        expect(password.getElementsByTagName('input')[0].getAttribute('type')).toBe("password");
    });

    it('renders submit input correctly', () => {
        const {getByTestId} = render(<Login/>);
        const submit = getByTestId('submit');
        expect(submit).toHaveValue("Submit");
        expect(submit.getAttribute('type')).toBe("submit");
    });

    it('renders cancel input correctly', () => {
        const {getByTestId} = render(<Login/>);
        const cancel = getByTestId('cancel');
        expect(cancel).toHaveValue("Cancel");
        expect(cancel.getAttribute('type')).toBe("button");
    });

    it('renders link to create account correctly', () => {
        const {getByTestId} = render(<Login/>);
        const createAccount = getByTestId('createAccount');
        //TODO -set link for createAccount page
        expect(createAccount.getAttribute('href')).toBe("#");
    });


    it('matches snapshot', () => {
        const tree = renderer.create(<Login/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
