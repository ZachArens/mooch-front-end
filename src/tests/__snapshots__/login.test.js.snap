// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Login component correctly matches snapshot 1`] = `
<div
  data-testid="login"
>
  <div
    hidden={true}
    id="loginError"
  >
    <h2>
      
    </h2>
  </div>
  <form>
    <title>
      Login
    </title>
    <label
      data-testid="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="text"
      />
    </label>
    <label
      data-testid="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="********"
        type="password"
      />
    </label>
    <div
      className="submitButtons"
    >
      <input
        data-testid="submit"
        onClick={[Function]}
        type="submit"
        value="Submit"
      />
      <input
        data-testid="cancel"
        type="button"
        value="Cancel"
      />
    </div>
    <a
      href="https://www.google.com"
    >
      <p>
        Don't have an account?  Create one.
      </p>
    </a>
  </form>
</div>
`;
